File: python/monty.py
import random

def experiment(n) -> int:
    win_count = 0

    for i in range(n):
        prize = random.randint(0, 2)
        choice = random.randint(0, 2)
        if choice == prize: 
            win_count += 1
    
    return win_count

def experiment2(n) -> int:
    win_count = 0
    
    for i in range(n):
        prize = random.randint(0, 2)
        choice = random.randint(0, 2)
        
        possible = []

        if choice == prize:
            possible = [choice, *filter(lambda x: x != choice, [0,1,2])]
        else: 
            possible = [prize, choice]

        choice = random.choices(possible)[0]
        
        if choice == prize:
            win_count += 1
            

    return win_count

if __name__ == "__main__":
    count = 1000000
    win_count = experiment(count)
    print(f"experiment #1 probability: {win_count * 100 / count}")
    win_count2 = experiment2(count)
    print(f"experiment #2 probability: {win_count2 * 100 / count}")

File: typescript/monty.js
var N = 1000000;
var winCount = Array.from(Array(N).keys())
    .map(function (_) { return ({
    choice: Math.floor(Math.random() * 3),
    prize: Math.floor(Math.random() * 3)
}); })
    .filter(function (x) { return x.choice == x.prize; })
    .length;
var winCount2 = Array.from(Array(N).keys())
    .map(function (_) { return ({
    choice: Math.floor(Math.random() * 3),
    prize: Math.floor(Math.random() * 3)
}); })
    .map(function (x) { return ({
    prize: x.prize,
    choice: (x.prize == x.choice
        ? [0, 1, 2][Math.floor(Math.random() * 3)]
        : [x.choice, x.prize][Math.floor(Math.random() * 2)])
}); })
    .filter(function (x) { return x.choice == x.prize; })
    .length;
console.log("experiment #1 probability: ".concat(winCount * 100 / N));
console.log("experiment #2 probability: ".concat(winCount2 * 100 / N));
